File: common/src/main/java/dev/obscuria/tooltips/client/element/icon/AccentIcon.java
package dev.obscuria.tooltips.client.element.icon;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import dev.obscuria.fragmentum.util.easing.Easing;
import dev.obscuria.tooltips.client.element.Transform;
import dev.obscuria.tooltips.client.renderer.TooltipContext;
import net.minecraft.client.gui.GuiGraphics;

public record AccentIcon(Transform transform) implements TooltipIcon {

    public static final Codec<AccentIcon> CODEC;

    @Override
    public Codec<AccentIcon> codec() {
        return CODEC;
    }

    @Override
    public void render(GuiGraphics graphics, TooltipContext context, int x, int y) {
        pushTransform(transform, graphics, context, x, y);
        graphics.renderItem(context.stack(), 0, 0);
        popTransform(graphics);
    }

    @Override
    public void applyScale(GuiGraphics graphics, TooltipContext context, int x, int y) {
        final var time = context.timeInSeconds();
        final var scale = (time < 0.25f)
                ? Easing.EASE_OUT_CUBIC.compute(time / 0.25f) * 1.25f
                : (time < 0.5f)
                ? 1.25f - 0.25f * Easing.EASE_OUT_CUBIC.compute((time - 0.25f) / 0.25f)
                : 1f;
        graphics.pose().scale(scale, scale, scale);
    }

    static {
        CODEC = RecordCodecBuilder.create(codec -> codec.group(
                Transform.CODEC.fieldOf("transform").forGetter(AccentIcon::transform)
        ).apply(codec, AccentIcon::new));
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/client/element/icon/AccentIcon.kt
package dev.obscuria.tooltips.client.element.icon

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import dev.obscuria.fragmentum.tools.easing.Easing
import dev.obscuria.tooltips.client.renderer.TooltipContext
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.world.phys.Vec3
import java.util.*

class AccentIcon(
    offset: Optional<Vec3>,
    scale: Optional<Float>
) : TransformableIcon(offset, scale) {

    override fun codec(): Codec<out ITooltipIcon> = CODEC

    override fun applyScale(graphics: GuiGraphics, context: TooltipContext, x: Int, y: Int) {

        val time = context.timeInSeconds()
        val scale = when {
            time < 0.25f -> Easing.EASE_OUT_CUBIC.compute(time / 0.25f) * 1.25f
            time < 0.5f -> 1.25f - 0.25f * Easing.EASE_OUT_CUBIC.compute((time - 0.25f) / 0.25f)
            else -> 1f
        }
        graphics.pose().scale(scale, scale, scale)
    }

    companion object {

        val CODEC: Codec<AccentIcon> = RecordCodecBuilder.create { codec ->
            codec.group(
                Vec3.CODEC
                    .optionalFieldOf("offset")
                    .forGetter(AccentIcon::offset),
                Codec.FLOAT
                    .optionalFieldOf("scale")
                    .forGetter(AccentIcon::scale)
            ).apply(codec, ::AccentIcon)
        }
    }
}